### ISSUE TITLE
continuous update rejects could flood envoy with discovery requests

### ISSUE THREAD

**Issue by ramaraochavali** [envoyproxy/envoy#4718]
If for a particular reason, an update is rejected, Envoy sends back a response immediately indicating the failure to management server. However if management server, responds again with incorrect data with possibility of update being rejected, Envoy would send the update rejected response immediately.
This continuous flow of rejections can be quite chatty and could flood envoy with discovery requests.

It is better to implement back-off for update rejected cases with back-off starting immediately after the first failure.

**Comment by htuch** [comment]
Arguably the management server should implement back-off here @ramaraochavali.

**Comment by ramaraochavali** [comment]
@htuch I was also thinking that management server should implement the back-off here. However following (especially the first one) was the motivation for the issue and the PR.
- Should Envoy protect itself from a management server misbehaviour possibly because of a bug in it's implementation or some other conditions?
- Would it be useful for all management servers if Envoy implements this instead of every one doing it?
I am Ok if we say it is a conscious design choice that management server should take care of this.

**Comment by ramaraochavali** [comment]
Another point to consider here request back-off vs. response back-off which one is better as a design choice? I tried both with this PR in envoy and implementing it in our management server on the response side - request side looked more natural to me, but it's just me. 
LMK what you think?

**Comment by junr03** [comment]
@htuch jumping in from the left field here. ~I vaguely remember someone bringing up a similar line of conversation a few months ago. I think we discussed "ratelimiting" discovery interactions back then.~ Found the discussion in #2169 (if you are interested, @ramaraochavali).

**Comment by htuch** [comment]
Yeah, I think we can do back-off in Envoy, we just need to make sure it doesn't affect normal rapid successful updates. I still think Envoy doesn't need to protect itself as the management server is essentially trusted (a broken management server can do a lot more bad things than cause Envoy to spin..). But, if you want to add this, seems legit.
